### PR TITLE
Selectable bounds/selection box refactor

### LINKED ISSUE #5552: Concrete walls cant be sold beside buildings
If you have a barrier behind a building you have to sell the building before you can sell the barrier.
concrete wall can t be sold 1

### PR DESCRIPTION
**Pull request by reaperrr**
- Introduced `VisualBounds` in `SelectionDecorations`, these are responsible for the visible selection box now (independent from actual selectable area)
- Moved `SelectionDecorations` and `SelectionBoxRenderable` to Mods.Common.
- selection boxes and selection bars can now be enabled/disabled individually, selection box color is now configurable.
- Set tighter `Selectable.Bounds` for many actors, to dramatically reduce issues where it was difficult to click on a cell or actor hidden behind another actor's selectable area.
- Removed `Selectable` boolean from `Selectable` trait.

This should pretty much fix #5552 and #3871 for the shipping mods. TS buildings will need isometric bounds to fix this completely.

### CODE REVIEW

**Comment by deniz1a** [comment]
Does this patch work by making selection areas smaller so they don't overlap? I found the following issue: When you build a turret on one of the edge cells of a service depot, the turret is not reachable. When the turret is targeted, the service depot is attacked.

**Comment by Mailaender** [comment]
Needs a rebase.

**Comment by reaperrr** [comment]
> When you build a turret on one of the corner cells of a service depot, the turret is sometimes not selectable. Also that cell is unreachable by units, as opposed to the ones above refinery.

Assuming you meant the RA service depot, I forgot to tweak its bounds.
Fixed.

@penev92:
- Moved debug path rendering to ```SelectionDecorations```
- used ```HasTrait``` to check for ```Selectable``` in ```WorldRenderer``` + ```UnitOrderGenerator```
- ```Selectable: true``` boolean is now removed as well (but not the trait in this case)
- Added console message if trait has been removed
- removed ```BodyOrientation``` + render traits in favor of new ```CustomSelectionSize```

> Needs a rebase.

Done.

**Comment by deniz1a** [comment]
The issue with ra service depot doesn't seem to be fixed. You have to click the edge of the turret, otherwise service depot is selected.

**Comment by reaperrr** [comment]
I just tested it myself again, for me at most the lower 3-4 pixels of the turret are overlapped by the service depot's selectable area. 
Try comparing with the service depot in TD mod, if you see a significant difference, check whether something might have gone wrong while fetching the updates.

**Comment by reaperrr** [comment]
Updated. Tweaked RA and TD service depot ```Bounds```, fixed TD ```VisualBounds``` (were mistyped as just ```Bounds```).

**Comment by deniz1a** [comment]
Ok now turrets at corners of service depot are selectable.

File: OpenRA.Game/Traits/Selectable.cs [hunk 1]
```diff
@@ -9,58 +9,18 @@
 #endregion
 
 using System.Collections.Generic;
-using System.Drawing;
 using System.Linq;
-using OpenRA.Graphics;
 
 namespace OpenRA.Traits
 {
 	public class SelectableInfo : ITraitInfo
```
Comment (Mailaender) [inline OpenRA.Game/Traits/Selectable.cs]: You can make shorten this to `TraitInfo<Selectable>` if this is otherwise empty.

File: OpenRA.Mods.Common/Traits/SelectionDecorations.cs [hunk 2]
```diff
@@ -9,39 +9,77 @@
...
-namespace OpenRA.Traits
+namespace OpenRA.Mods.Common.Traits
 {
-	public class SelectionDecorationsInfo : ITraitInfo
+	public class SelectionDecorationsInfo : ITraitInfo, ISelectionDecorationsInfo
 	{
 		public readonly string Palette = "chrome";
+		[Desc("Visual bounds for selection box. If null, it uses AutoSelectionSize.")]
+		public readonly int[] VisualBounds = null;
+		[Desc("Health bar, production progress bar etc.")]
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/SelectionDecorations.cs]: Newlines are often helpful so you can see at first glance which Desc belongs to which fields.

**Comment by penev92** [comment]
I'm not even convinced the corner cells of the service depots should be buildable on. The service depot artwork sure doesn't suggest it.

**Comment by reaperrr** [comment]
> I'm not even convinced the corner cells of the service depots should be buildable on. The service depot artwork sure doesn't suggest it.

To be fair, it was possible on the upper corners in the originals. Not on the lower ones due to bib, though.

**Comment by reaperrr** [comment]
Updated.

**Comment by reaperrr** [comment]
Rebased.

> This isn't ```additional``` IMO, this is a standard part of selection logic.

Removed ```Additionally``` from ```Selectable``` description.

File: OpenRA.Mods.Common/Traits/CustomSelectionSize.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Special case trait for invisible, unselectable actors like RA bridge huts.",
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/CustomSelectionSize.cs]: I would say we should avoid mod specific comments in trait descriptions especially common ones.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/CustomSelectionSize.cs]: All the mods except D2K have bridge huts, so just drop 'RA'.

**Comment by reaperrr** [comment]
Updated & rebased.

**Comment by abcdefg30** [comment]
Needs a rebase again.

**Comment by reaperrr** [comment]
Rebased.

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs [hunk 4]
```diff
@@ -305,8 +305,7 @@ static IEnumerable<Actor> SelectActorsInBoxWithDeadzone(World world, int2 a, int
 			return world.ScreenMap.ActorsInBox(a, b)
 				.Where(x =>
 				{
-					var s = x.TraitOrDefault<Selectable>();
-					return s != null && s.Info.Selectable && (x.Owner.IsAlliedWith(world.RenderPlayer) || !world.FogObscures(x));
+					return x.HasTrait<Selectable>() && (x.Owner.IsAlliedWith(world.RenderPlayer) || !world.FogObscures(x));
```
Comment (matija-hustic) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: This can now be further simplified to
`x => x.HasTrait<Selectable>() && (x.Owner.IsAlliedWith(world.RenderPlayer) || !world.FogObscures(x))`

Comment (penev92) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: A.k.a. remove the `return` and the `;` ;)

**Comment by reaperrr** [comment]
Updated.

**Comment by Mailaender** [comment]
TS crates seem to have lost their tooltips.

**Comment by Mailaender** [comment]
Also the TD crates don't show a tooltip. Not sure if that is even a regression or done on purpose. Everything is still selectable. :white_check_mark:

**Comment by pchote** [comment]
> Also the TD crates don't show a tooltip. 

This is a regression, and should be fixed.

**Comment by reaperrr** [comment]
Updated, crate tooltips should work again.

File: OpenRA.Mods.Common/Traits/SelectionDecorations.cs [hunk 5]
```diff
@@ -9,39 +9,81 @@
...
 
-namespace OpenRA.Traits
+namespace OpenRA.Mods.Common.Traits
 {
-	public class SelectionDecorationsInfo : ITraitInfo
+	public class SelectionDecorationsInfo : ITraitInfo, ISelectionDecorationsInfo
 	{
 		public readonly string Palette = "chrome";
 
+		[Desc("Visual bounds for selection box. If null, it uses AutoSelectionSize.")]
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/SelectionDecorations.cs]: Please add to the description that the first two elements are the bounds' size, while *optional* third and forth elements are offset that affects the location of the selection decorations.

Comment (penev92) [inline OpenRA.Mods.Common/Traits/SelectionDecorations.cs]: Also I think it should default to `Selectable.Bounds` (might even save some work in `Actor`'s constructor).

File: OpenRA.Game/Actor.cs [hunk 6]
```diff
@@ -110,6 +112,19 @@ internal Actor(World world, string name, TypeDictionary initDict)
 				return new Rectangle(offset.X, offset.Y, size.X, size.Y);
 			});
 
+			visualBounds = Exts.Lazy(() =>
+			{
+				var sd = Info.Traits.GetOrDefault<ISelectionDecorationsInfo>();
```
Comment (penev92) [inline OpenRA.Game/Actor.cs]: I think it should default to `Selectable.Bounds` if there isn't a trait that implements `ISelectionDecorationsInfo`, or it doesn't specify visual bounds.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 7]
```diff
@@ -1149,6 +1149,51 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
 					}
 				}
 
+				if (engineVersion < 20150603)
```
Comment (penev92) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: It's always nice to leave a comment on each upgrade rule to say what it's about. Makes navigating this huge file much easier.

Comment (penev92) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: You should probably also update the dates on both rules.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 8]
```diff
@@ -1149,6 +1149,51 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
...
+						var selectable = node.Value.Nodes.FirstOrDefault(n => n.Key == "Selectable");
+						var selDecor = node.Value.Nodes.FirstOrDefault(n => n.Key == "SelectionDecorations");
+						var selectableNodes = selectable.Value.Nodes;
+						var bounds = selectableNodes.FirstOrDefault(n => n.Key == "Bounds");
+
+						if (bounds != null)
+						{
+							var visualBounds = FieldLoader.GetValue<string>("Bounds", bounds.Value.Value);
+							if (selDecor != null)
+								selDecor.Value.Nodes.Add(new MiniYamlNode("VisualBounds", visualBounds.ToString()));
```
Comment (penev92) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Redundant `.ToString()` (visualBounds is already a string).

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 9]
```diff
@@ -438,6 +440,7 @@
     <Compile Include="Traits\RepairsBridges.cs" />
     <Compile Include="Traits\SeedsResource.cs" />
     <Compile Include="Traits\StoresResources.cs" />
+    <Compile Include="Traits\SelectionDecorations.cs" />
```
Comment (penev92) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: How about `Traits\Render\SelectionDecorations.cs` ?

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 10]
```diff
@@ -110,6 +110,16 @@ public interface IDisableMove { bool MoveDisabled(Actor self); }
 
 	public interface ISeedableResource { void Seed(Actor self); }
 
+	public interface ISelectionDecorations
```
Comment (penev92) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]: I don't think this is at all necessary.

File: OpenRA.Game/Traits/Selectable.cs [hunk 11]
```diff
@@ -9,16 +9,16 @@
 #endregion
 
 using System.Collections.Generic;
-using System.Drawing;
 using System.Linq;
-using OpenRA.Graphics;
 
 namespace OpenRA.Traits
 {
+	[Desc("This actor is selectable. Defines bounds of selectable area and selection priority.")]
```
Comment (penev92) [inline OpenRA.Game/Traits/Selectable.cs]: Now also selection class.

File: OpenRA.Mods.Common/Traits/CustomSelectionSize.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+ */
+#endregion
+
+using System;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Special case trait for invisible, unselectable actors like bridge huts.",
+	"Gives actor targetable area for special cases like C4 and engineer repair.",
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/CustomSelectionSize.cs]: How about
"Special case trait for unselectable actors that need do define targetable area bounds for special cases like C4, engineer repair and tooltips. Examples: bridge huts and crates."

File: OpenRA.Game/Traits/Selectable.cs [hunk 13]
```diff
@@ -9,16 +9,16 @@
 #endregion
 
 using System.Collections.Generic;
-using System.Drawing;
 using System.Linq;
```
Comment (penev92) [inline OpenRA.Game/Traits/Selectable.cs]: Minor nit: both of these two remaining `using`s are now unnecessary as well.

File: OpenRA.Game/Traits/Selectable.cs [hunk 14]
```diff
@@ -28,46 +28,16 @@ public class SelectableInfo : ITraitInfo
...
 	}
 
-	public class Selectable : IPostRenderSelection
+	public class Selectable
 	{
 		public readonly string Class = null;
 
-		public SelectableInfo Info;
-		readonly Actor self;
+		public readonly SelectableInfo Info;
```
Comment (penev92) [inline OpenRA.Game/Traits/Selectable.cs]: Minor nit: This no longer needs to be `public`.

**Comment by penev92** [comment]
OK, I *may* have overdone it a bit, but I think everything I requested is simple. I'll test it ingame later, hopefully after you've had the chance to address my comments.

**Comment by reaperrr** [comment]
Updated (and rebased). The last 6 commits are new.

I didn't feel like cleaning up all ```VisualBounds``` in yaml even though quite a few of them are now redundant, since ```VisualBounds``` now defaults to ```Selectable.Bounds``` if ```null```.

Tell me if some commits should be squashed, I just wanted to keep the latest changes separate for now.

**Comment by penev92** [comment]
Thanks for not squashing :P
I'll review the new commits and then try and work out what should be squashed where.

File: OpenRA.Game/Actor.cs [hunk 15]
```diff
@@ -41,12 +41,23 @@ public sealed class Actor : IScriptBindable, IScriptNotifyBind, ILuaTableBinding
...
 		Lazy<IEffectiveOwner> effectiveOwner;
 
 		public Rectangle Bounds { get { return bounds.Value; } }
+
+		public Rectangle VisualBounds
+		{
+			get
+			{
+				var sd = Info.Traits.GetOrDefault<ISelectionDecorationsInfo>();
+				return (sd != null && sd.SelectionBoxBounds != null) ? visualBounds.Value : bounds.Value;
```
Comment (penev92) [inline OpenRA.Game/Actor.cs]: Sorry, this will look of traits each time you select an actor, for each actor.
What I meant was check in `visualBounds` ([here](https://github.com/OpenRA/OpenRA/pull/8388/files#diff-25e3f19d154f93bcb747e9f12c252028R126)) and if there is no such trait, give it `bounds`' value (or make it `null` and then do `if (visualBounds == null) return bounds.Value;` here).

File: OpenRA.Game/Actor.cs [hunk 16]
```diff
@@ -110,6 +121,19 @@ internal Actor(World world, string name, TypeDictionary initDict)
 				return new Rectangle(offset.X, offset.Y, size.X, size.Y);
 			});
 
+			visualBounds = Exts.Lazy(() =>
```
Comment (penev92) [inline OpenRA.Game/Actor.cs]: My suggestion here is as follows:
```
			visualBounds = Exts.Lazy(() =>
			{
				var sd = Info.Traits.GetOrDefault<ISelectionDecorationsInfo>();
				if (sd == null || sd.SelectionBoxBounds == null)
					return bounds.Value;

				var size = new int2(sd.SelectionBoxBounds[0], sd.SelectionBoxBounds[1]);

				var offset = -size / 2;
				if (sd.SelectionBoxBounds.Length > 2)
					offset += new int2(sd.SelectionBoxBounds[2], sd.SelectionBoxBounds[3]);

				return new Rectangle(offset.X, offset.Y, size.X, size.Y);
			});
```

**Comment by penev92** [comment]
OK with that slight change, I think I can +1 now.
Also don't forget to update the date on the upgrade rule (again).

File: OpenRA.Game/Traits/Selectable.cs [hunk 17]
```diff
@@ -28,46 +25,16 @@ public class SelectableInfo : ITraitInfo
...
 	}
 
-	public class Selectable : IPostRenderSelection
+	public class Selectable
 	{
 		public readonly string Class = null;
 
-		public SelectableInfo Info;
-		readonly Actor self;
+		readonly SelectableInfo info;
```
Comment (penev92) [inline OpenRA.Game/Traits/Selectable.cs]: Travis is failing because this is unused. Just remove it.

File: OpenRA.Game/Actor.cs (hunk 15 above)
Comment (reaperrr) [inline OpenRA.Game/Actor.cs]: > Edit: Rectangle apparently can't be null.

Yeah, I tried that at first.

**Comment by reaperrr** [comment]
Updated.

**Comment by penev92** [comment]
Great!
:+1: 
Thanks for another nice cleanup :)

**Comment by reaperrr** [comment]
No problem. Should I squash something now, or can this be merged as-is?

**Comment by penev92** [comment]
Usually I'd like them squashed, but ... meh.

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/553866e8b687e7db98858dda5cc991767d9f4053...5bb676140b5dac0ec4e1e5f07afe5f118ed6548d)
That's how far my imagination took me. If you think I missed something, feel free to add it, or to generally make the log better.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 18]
```diff
@@ -1152,6 +1152,29 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
...
+							node.Key = "SelectableRemoveMe";
+
+						// To cover rare cases where the boolean was 'true'
+						if (node.Key == "Selectable" && selectable.Value.Value == "true")
+							node.Value.Nodes.Remove(selectable);
+					}
+
+					if (depth == 0 && node.Value.Nodes.Exists(n => n.Key == "SelectableRemoveMe"))
+						node.Value.Nodes.RemoveAll(n => n.Key == "SelectableRemoveMe");
+						Console.WriteLine("The 'Selectable' boolean has been removed from the Selectable trait.");
```
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: You forgot the braces here. Found out by accident when MonoDevelop corrected the indentions. Fixed it in https://github.com/OpenRA/OpenRA/pull/8412.

File: OpenRA.Game/Traits/Selectable.cs (hunk 17 above)
Comment (Mailaender) [inline OpenRA.Game/Traits/Selectable.cs]: I reverted it in https://github.com/OpenRA/OpenRA/pull/8412.
